### DISCUSSION TITLE
Copilot much worse then free ChatGPT 3.5

### DISCUSSION THREAD

**Discussion by fbucek** [community/community#69998] [Copilot Conversations] [answered]
### Copilot

Product Feedback

### Copilot much worse then free ChatGPT 3.5

Whenever I am trying to use copilot feature like chat it gives me much worse responses then free ChatGPT.

Moreover, copilot shoud have more context, but it does not solve sometime primitive questions which ChatGPT solve like master. 

Example which I have created for this.
```rust
pub fn get_path() -> &str {
    // check if running on macos
    if cfg!(target_os = "macos") {
       PATH_MAC
    } else {
       PATH_SERVER
    }
}
``` 

Copilot chat: fix this -> response unrecognized language: rust 
Copilot chat: fix this in rust programming language

Result: It did not fix, but copied PATH_MAC, PATH_SERVER constatns above this method, keeping everything as before, so ite made it worse. Error was not solve and it duplicated code which cuased another two errors. But this proves that even small part is selected it has access to whole file. So it has more context to help.

ChatGPT: fix this: ( and I have pasted just  code in example no context ) 
Resutl: It fixed it and made &str -> &'static str. So it was correct fix.  Then I asked to make it 'a lifetime -> it correctly fixed to 'a lifetime so it also added <'a> part to method signiture.

Summary:
Copilot has much more context to solve problem, but is much worse then ChatGPT with little context. Same goes for another languages, like typescript, python which I use. It is slightly better in this languages, but problem is the same. Much worse results. 
I have not reason to press CMD+I to ask question, when I know result will be much worse, then using free chatgpt.
I do not remember what copilot chat helped me to solve anything at all.

Code suggestion helps a lot, so when I am writing in multiple languages it always uses correct syntax, that's very good.
But copilot chat is really terrible same goes to all the brushes in VSCode extension, which has also thight limitiation. Such a unitests only for js and ts, where in free chatgpt create unitestest almost for any language by writing: create unittest. And it correctly creates with filled example data.

I would understand when it would be in reverse free chatGPT with much worse results then 100$ copilot, but it is not.

**Comment by kuroodo** [comment, marked as answer]
Copilot should be used as just that: a copilot. I believe it still uses the Codex model rather than GPT-3.5 or GPT-4, which is why its so much worse. The best way to make the most out of copilot is to allow it to autocomplete code for you, and generate code while coding. If you need a quick explanation, recommendation, or anything else that requires the chat, while it is more convenient and quicker to use the chat itself since it already has a context to your code, you will always get a much better response from GPT-3.5 and especially GPT-4. Overall copilot chat has been an extremely frustrating experience for me.

There's a free alternative to copilot called Codeium that you should try out. It also has a chat interface (had it before copilot) and while it still is not as good as GPT, I found it to produce much better and more accurate responses compared to copilot. As for the other aspects, I found it to generally perform on the same level as copilot. Haven't tried unit tests with it though. But the team at Codeium apparently retrains their model every 2 months or so, and they are SOC2 compliant unlike copilot. Could be a good alternative so you can shell out $20 for GPT-4 access with Codeium, rather than $20 for GPT-4 + $10 for copilot.

**Reply by fbucek** [reply]
Thanks, codeium looks very promising.

**Reply by SaidTorres3** [reply]
Codeium is better than copilot for me, it has better coding.

**Reply by arthabus** [reply]
I was surprised to learn myself that copilot performs poorly in this task compared to chatgpt, since it's supposed to be trained on code and intuitively should be better than a general tool, which is chatgpt in this context. Obviously this is more complex than that. Gonna try codeium.

> Copilot should be used as just that: a copilot.

I'd say your definition of "a copilot" as solely code completion is rather debatable. 

They've added chat, clearly this falls under "copiloting" in the view of creators. And that makes perfect sense to me too - I need assistance in development and I don't care if it's code completion or if it's a chat. I wouldn't narrow down this definition.
